Improve freopen

pfusik opened the pull request.

pfusik commented:
Fixes #354 and #356

mysterymath commented on mos-platform/common/c/stdio-full.c:
```diff
@@ -128,36 +128,7 @@ static unsigned filemode(const char *const mode) {
-}
-
-static signed char stdio_open(const char *const filename, unsigned int mode) {
+static FILE *stdio_open(FILE *stream, const char *filename, unsigned int mode) {
```
This needs a comment describing the relationship between the argument and the return value. In particular, it wasn't clear that this takes ownership of the stream argument, and that that argument must be dynamically allocated.

pfusik replied:
Done.
`stream` is not necessarily `malloc`ed - it can be `stdout` from `freopen`. The old code was doing `free` without checking that - this is one of the bugs I'm fixing.

mysterymath commented on mos-platform/common/c/stdio-full.c:
```diff
@@ -292,8 +281,13 @@ int fclose(FILE *stream) {
-  return 0;
+int fclose(FILE *stream) {
+  int rc = stdio_close(stream);
+
```
nit: delete line.

pfusik replied:
Done.

mysterymath commented on mos-platform/common/c/stdio-full.c:
```diff
@@ -87,6 +85,86 @@ void _stdio_closeall(void) {
+
+static void free_stream(FILE *stream) {
+  /* Remove stream from list */
+  if (filelist == stream)
```
nit: One aspect of the LLVM style I'd like to keep is that if one clause of an if/else has braces; the other must also. Making them the two clauses visually similar tends to help with finding their ranges at a glance.

pfusik replied:
Done.

mysterymath commented on mos-platform/common/c/stdio-full.c:
```diff
@@ -367,110 +335,20 @@ FILE *freopen(const char *restrict filename, const char *restrict mode,
-      return NULL;
-
-    case 0:
-      /* fail; try close / reopen */
```
This PR should have `[Breaking]` in its description, since this is a change to existing implementation-defined behavior.

mysterymath replied:
Since we're rebasing; this should be done on the individual commit that introduces the breaking change.

pfusik replied:
Please read the new commit messages. If we split the first commit, the motivation could be become unclear.

mysterymath reviewed (commented):
This PR description needs an itemized summary of the contained changes; otherwise it will be difficult to write the release notes post-squash. Particularly since it's a breaking change.

pfusik commented:
Do we need to squash this? These are three related but separate changes. I would prefer to squash only the fixup commit that addresses review comments.

mysterymath commented:
> Do we need to squash this? These are three related but separate changes. I would prefer to squash only the fixup commit that addresses review comments.

I suppose we can do that.

mysterymath reviewed (approved):
LGTM, thanks!
I'll hold off on committing until the squash; let me know.

pfusik commented on mos-platform/common/c/stdio-full.c:
```diff
@@ -208,6 +208,10 @@ static unsigned filemode(const char *const mode) {
 
+// Initializes stream by opening the handle
+// and initializing most of structure members.
+// stream->buffer, stream->bufsize and stream-next must be initialized on entry.
```
Fixing `stream-next` typo while squashing.
